File: src/types/_ActivatableSkillRange.ts
import * as DB from "tsondb/schema/dsl"
import { CheckResultBasedModifier, CheckResultValue } from "./_ActivatableSkillCheckResultBased.js"
import { SkillModificationLevelIdentifier } from "./_Identifier.js"
import { ResponsiveTextOptional, ResponsiveTextReplace } from "./_ResponsiveText.js"
import { NestedTranslationMap } from "./Locale.js"

export const Range = DB.TypeAlias(import.meta.url, {
  name: "Range",
  type: () =>
    DB.Object({
      value: DB.Required({
        comment: "The range value.",
        type: DB.IncludeIdentifier(RangeValue),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "Range",
        DB.Object(
          {
            note: DB.Optional({
              comment:
                "A note, appended to the generated string in parenthesis. If the generated is modified using `replacement`, the note is appended to the modifier string.",
              type: DB.IncludeIdentifier(ResponsiveTextOptional),
            }),
            replacement: DB.Optional({
              comment:
                "A replacement string. If `note` is provided, it is appended to the replaced string.",
              type: DB.IncludeIdentifier(ResponsiveTextReplace),
            }),
          },
          { minProperties: 1 },
        ),
      ),
    }),
})

const RangeValue = DB.Enum(import.meta.url, {
  name: "RangeValue",
  values: () => ({
    Modifiable: DB.EnumCase({ type: DB.IncludeIdentifier(ModifiableRange) }),
    Sight: DB.EnumCase({ type: null }),
    Self: DB.EnumCase({ type: null }),
    Global: DB.EnumCase({ comment: "German: *dereumfassend*", type: null }),
    Touch: DB.EnumCase({ type: null }),
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FixedRange) }),
    CheckResultBased: DB.EnumCase({ type: DB.IncludeIdentifier(CheckResultBasedRange) }),
  }),
})

const ModifiableRange = DB.TypeAlias(import.meta.url, {
  name: "ModifiableRange",
  type: () =>
    DB.Object({
      is_maximum: DB.Optional({
        comment: "If `true`, the range is a maximum range.",
        type: DB.Boolean(),
      }),
      initial_modification_level: DB.Required({
        comment: "The initial skill modification identifier/level.",
        type: SkillModificationLevelIdentifier(),
      }),
      is_radius: DB.Optional({
        comment: "If `true`, the range is a radius.",
        type: DB.Boolean(),
      }),
    }),
})

export const FixedRange = DB.TypeAlias(import.meta.url, {
  name: "FixedRange",
  type: () =>
    DB.Object({
      is_maximum: DB.Optional({
        comment: "If `true`, the range is a maximum range.",
        type: DB.Boolean(),
      }),
      value: DB.Required({
        comment: "The (unitless) range value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      unit: DB.Required({
        comment: "The unit of the `value`.",
        type: DB.IncludeIdentifier(RangeUnit),
      }),
      is_radius: DB.Optional({
        comment: "If `true`, the range is a radius.",
        type: DB.Boolean(),
      }),
    }),
})

const CheckResultBasedRange = DB.TypeAlias(import.meta.url, {
  name: "CheckResultBasedRange",
  type: () =>
    DB.Object({
      is_maximum: DB.Optional({
        comment: "If the range is the maximum range.",
        type: DB.Boolean(),
      }),
      base: DB.Required({
        comment: "The base value that is derived from the check result.",
        type: DB.IncludeIdentifier(CheckResultValue),
      }),
      modifier: DB.Optional({
        comment: "If defined, it modifies the base value.",
        type: DB.IncludeIdentifier(CheckResultBasedModifier),
      }),
      unit: DB.Required({
        comment: "The duration unit.",
        type: DB.IncludeIdentifier(RangeUnit),
      }),
      is_radius: DB.Optional({
        comment: "If `true`, the range is a radius.",
        type: DB.Boolean(),
      }),
    }),
})

const RangeUnit = DB.Enum(import.meta.url, {
  name: "RangeUnit",
  values: () => ({
    Steps: DB.EnumCase({ type: null }),
    Miles: DB.EnumCase({ type: null }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/_ActivatableSkillRange.swift
import FileDB

@Embedded
public struct Range {

  /// The range value.
  let value: RangeValue

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]?

    @Embedded
    @MinProperties(1)
    struct Translation { // RangeTranslation
            /// A note, appended to the generated string in parenthesis. If the generated is modified using `replacement`, the note is appended to the modifier string.
            let note: ResponsiveTextOptional?

            /// A replacement string. If `note` is provided, it is appended to the replaced string.
            let replacement: ResponsiveTextReplace?
          }
  }

@ModelEnum
public enum RangeValue {
    case Modifiable(ModifiableRange)
    case Sight
    case `Self`
    /// German: *dereumfassend*
    case Global
    case Touch
    case Fixed(FixedRange)
    case CheckResultBased(CheckResultBasedRange)
}

@Embedded
public struct ModifiableRange {
  /// If `true`, the range is a maximum range.
  let is_maximum: Bool?

  /// The initial skill modification identifier/level.
  @Relationship(SkillModificationLevel.self)
  let initial_modification_level: SkillModificationLevel.ID

  /// If `true`, the range is a radius.
  let is_radius: Bool?
  }

@Embedded
public struct FixedRange {
  /// If `true`, the range is a maximum range.
  let is_maximum: Bool?

  /// The (unitless) range value.
  @Minimum(1)
  let value: Int

  /// The unit of the `value`.
  let unit: RangeUnit

  /// If `true`, the range is a radius.
  let is_radius: Bool?
  }

@Embedded
public struct CheckResultBasedRange {
  /// If the range is the maximum range.
  let is_maximum: Bool?

  /// The base value that is derived from the check result.
  let base: CheckResultValue

  /// If defined, it modifies the base value.
  let modifier: CheckResultBasedModifier?

  /// The duration unit.
  let unit: RangeUnit

  /// If `true`, the range is a radius.
  let is_radius: Bool?
  }

@ModelEnum
public enum RangeUnit {
    case Steps
    case Miles
}
